--- a/pptp/nss_connmgr_pptp.c
+++ b/pptp/nss_connmgr_pptp.c
@@ -146,7 +146,7 @@ static int nss_connmgr_pptp_get_session(
 	/*
 	 * check whether the interface is of type PPP
 	 */
-	if (dev->type != ARPHRD_PPP || !(dev->priv_flags & IFF_PPP_PPTP)) {
+	if (dev->type != ARPHRD_PPP || !(dev->priv_flags_qca_ecm & IFF_QCA_ECM_PPP_PPTP)) {
 		nss_connmgr_pptp_info("%p: netdevice is not a PPP tunnel type\n", dev);
 		return -1;
 	}
@@ -686,7 +686,7 @@ static int nss_connmgr_pptp_dev_down(str
 	/*
 	 * check whether the interface is of type PPP
 	 */
-	if (dev->type != ARPHRD_PPP || !(dev->priv_flags & IFF_PPP_PPTP)) {
+	if (dev->type != ARPHRD_PPP || !(dev->priv_flags_qca_ecm & IFF_QCA_ECM_PPP_PPTP)) {
 		nss_connmgr_pptp_info("%p: netdevice is not a pptp tunnel type\n", dev);
 		return NOTIFY_DONE;
 	}
