--- a/frontends/nss/ecm_nss_ipv4.c
+++ b/frontends/nss/ecm_nss_ipv4.c
@@ -2170,6 +2170,9 @@ sync_conntrack:
 				set_bit(IPS_ASSURED_BIT, &ct->status);
 
 				timeouts = nf_ct_timeout_lookup(ct);
+				if (!timeouts) {
+					timeouts = udp_get_timeouts(nf_ct_net(ct));
+				}
 
 				spin_lock_bh(&ct->lock);
 				ct->timeout = jiffies + timeouts[UDP_CT_REPLIED];
--- a/frontends/nss/ecm_nss_ipv6.c
+++ b/frontends/nss/ecm_nss_ipv6.c
@@ -1882,6 +1882,9 @@ sync_conntrack:
 				set_bit(IPS_ASSURED_BIT, &ct->status);
 
 				timeouts = nf_ct_timeout_lookup(ct);
+				if (!timeouts) {
+					timeouts = udp_get_timeouts(nf_ct_net(ct));
+				}
 
 				spin_lock_bh(&ct->lock);
 				ct->timeout = jiffies + timeouts[UDP_CT_REPLIED];
--- a/frontends/sfe/ecm_sfe_ipv4.c
+++ b/frontends/sfe/ecm_sfe_ipv4.c
@@ -1607,6 +1607,9 @@ sync_conntrack:
 				set_bit(IPS_ASSURED_BIT, &ct->status);
 
 				timeouts = nf_ct_timeout_lookup(ct);
+				if (!timeouts) {
+					timeouts = udp_get_timeouts(nf_ct_net(ct));
+				}
 
 				spin_lock_bh(&ct->lock);
 				ct->timeout = jiffies + timeouts[UDP_CT_REPLIED];
--- a/frontends/sfe/ecm_sfe_ipv6.c
+++ b/frontends/sfe/ecm_sfe_ipv6.c
@@ -1320,6 +1320,9 @@ sync_conntrack:
 				set_bit(IPS_ASSURED_BIT, &ct->status);
 
 				timeouts = nf_ct_timeout_lookup(ct);
+				if (!timeouts) {
+					timeouts = udp_get_timeouts(nf_ct_net(ct));
+				}
 
 				spin_lock_bh(&ct->lock);
 				ct->timeout = jiffies + timeouts[UDP_CT_REPLIED];
